### PR TITLE
node-migrate es2017 version featuring async/await

### PR DESCRIPTION
**Pull request by jdpnielsen**
Replaces callbacks with promises & allows for async/await & normalizes a bunch of stuff.
I feel this really simplifies complex migrations - take a look at tests & examples.

Has the unfortunate side effect of dropping support for older node versions, so this might have to live on in a fork, but let me know your thoughts.

### CODE REVIEW

**Comment by wesleytodd** [comment]
Hey @jdpnielsen!  Wow, thanks for all the work.  Can you describe, from your perspective, what the changes in this PR achieve?

> I feel this really simplifies complex migrations

By this comment you seem to be saying that the user experience of the tool is the goal.  But re-writing the internals doesn't seem to matter for that.  If your goal is for users to be able to write migrations with `async`/`await` and `Promise`'s, then a simple helper library or wrapper would achieve the same result.

Again, thanks for the contribution, and hopefully we can get something which solves your use case.

**Comment by jdpnielsen** [comment]
Initially my goal was to improve the user experience. I choose to go ahead and refactor the internals after seeing how it simplified everything. 

I could definitely use a wrapper instead but, in my opinion, it would add too much complexity on the user's end - or in the internals, depending on where such a wrapper would be implemented.

**Comment by wesleytodd** [comment]
I don't think there is a case were we would want to enforce using `async`/`await` for migrations, which is what this does (correct me if I am wrong).  Maybe we could integrate a way to support it?  I think adding support is a MUCH better approach.  Any ideas are welcome.

As for refactoring the internals, I am a fairly strong thumbs down on this.  I don't think it simplifies things, and since it does not really change the behavior, I think it falls into the "change for the sake of using new features" category.  Which I am almost always against.

**Comment by jdpnielsen** [comment]
No, the only requirement is to return a promise. The behavior changes in the internals was intentionally kept to a minimum, as to keep PR as slim as possible. The overhaul was meant to make it easier to maintain the internals.
I'll go ahead and close this PR and instead maintain a fork at https://github.com/jdpnielsen/node-migrate - ill change the name later & add a reference to node-migrate.

**Comment by jdpnielsen** [comment]
Took a stab at implementing promises here: https://github.com/tj/node-migrate/pull/82.

**Comment by sbolel** [comment]
This works really well with async/await

File: lib/template-generator.js [hunk 1]
```diff
@@ -1,51 +1,51 @@
...
+const path = require('path')
+const fs = require('fs')
+const util = require('util')
+const slug = require('slug')
+const formatDate = require('dateformat')
+const mkdirp = require('mkdirp')
+
+const pReadFile = util.promisify(fs.readFile)
+const pWriteFile = util.promisify(fs.writeFile)
+const pMkdirp = util.promisify(mkdirp)
```
Comment (sbolel) [inline lib/template-generator.js:12]: For some reason, with the newest versions of the dependencies, this was throwing an error. 

```
Template generation error : invalid options argument
```

I replaced the `util.promisify` call with the following:

```js
const pMkdirp = async (dir) => {
  const promise = new Promise((resolve, reject) => {
    return mkdirp(dir)
      .then(res => resolve(res))
      .catch(err => reject(err))
  })
  return await promise
}
```

And got it working. FYI.

Comment (jdpnielsen) [reply lib/template-generator.js:12]: Alternatively you could probably just use https://nodejs.org/dist/latest-v10.x/docs/api/fs.html#fs_fspromises_readfile_path_options

You are more than welcome to submit a PR to https://github.com/jdpnielsen/node-migrate

Further discussion should probably also be moved there.

Comment (sbolel) [reply lib/template-generator.js:12]: Thanks for the suggestion, I'll give it a shot. Started watching node-migrate as well and will help where I can!

**Pull request closed without merging**